jobserver/authorization/utils.py

```diff
@@ -0,0 +1,21 @@
+import itertools
+
+
+def has_permission(user, permission):
+    if not user.is_authenticated:
+        return False
+
+    roles = user.roles
+
+    # Each role has N permissions, combine them all into a final set
+    permissions = set(itertools.chain.from_iterable(r.permissions for r in roles))
```

Comment:
I would suggest a for loop and early return if found here. Constructing a full set for the likely common case of having the required permission is unnecessarily slow. And while it might not make much difference now, where its simple looping, down the line fetching all roles/permissions could get more expensive, e.g. with the local context stuff.

Another option might be caching this operation per request? Idk how often we'd expect permissions checks in a single request, but I'd imagine it could be quite a lot (e.g. templates checking to render cancel buttons for a list of jobs).

Reply:
I agree that early short-circuiting is probably sensible, as long as it's readable (`next(filter(pred, iterable), default)` is the way I have often seen of doing this but I don't like it!).

YAGNI on caching though, perhaps?

Reply:
Ugh, yes, -1 on `next(filter(pred, iterable))`. I think a simple for loop with if ....: return True is the most readable.

On caching, sure, I was just suggesting an alternative.

Reply:
Caching has definitely been on my mind, but we don't need it yet (YAGNI-yet if you will) and I'm keen to avoid performance optimisations until they're necessary.  Our page loads are currently in the 30-100ms range (SQLite 🎉).

Relatedly I don't see any performance problems with building the full set.  I've tried out a for-loop with a short-circuit

```
    permissions = itertools.chain.from_iterable(r.permissions for r in roles)
    for permission in permissions:
        if permission == permission:
            return True
```

and don't find that more readable than a membership test.

Reply:
Well, for a start you wouldn't need the itertools, as you can just nest for loops. That is more readable to me.

What happens when there's a project context to check though? If you are issuing DB queries for those, building the full set becomes more expensive, no?

And, if we are not caching, we could be checking a permission this 10s of times per request, building the set each time?

Just seems like an easy win to me :man_shrugging: